lib/Sema/CSSimplify.cpp

```diff
@@ -11499,6 +11499,67 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyFixConstraint(
       }
     }
 
+    auto anchor = locator->getAnchor();
+    auto *AE = getAsExpr<ApplyExpr>(anchor);
+    if (AE && getContextualTypePurpose(anchor) == CTP_ReturnSingleExpr) {
+      auto resultTy = getType(AE);
+      auto overload = findSelectedOverloadFor(AE->getSemanticFn());
+      auto resultTypeVar = resultTy->getAs<TypeVariableType>();
+      auto applyFnType = overload->openedType->castTo<FunctionType>();
+      // When a function result being returned is a generic parameter that can
+      // be inferred by a closure argument result e.g.
+      //   func callit<T>(_ f: () -> T) -> T {
+      //      f()
+      //   }
+      //
+      //   func caller() -> Int {
+      //      callit {
+      //        print("hello")
+      //      }
+      //   }
+      // let's score a contextual mismatch for the result call a bit higher
+      // to avoid ambiguity because we may have contextual mismatches for
+      // the closure as well.
+      if (overload && resultTypeVar &&
+          resultTypeVar->getImpl().getGenericParameter()) {
+        auto typeInvolvesTypeVar = [](Type type, TypeVariableType *typeVar) {
+          bool contains = false;
+          type.visit([&](Type ty) {
+            if (ty->isEqual(typeVar))
+              contains = true;
+          });
+          return contains;
+        };
+
+        // A single closure argument.
+        if (isExpr<ClosureExpr>(AE->getArg()->getSemanticsProvidingExpr())) {
+          auto param = applyFnType->getParams().front();
+          auto closureParamType =
+              param.getParameterType()->getAs<FunctionType>();
+          if (closureParamType &&
+              typeInvolvesTypeVar(closureParamType->getResult(),
```

Comment:
I think the type of the closure is supposed to be the same as result of the function being called since if it only contains it it's a completely difference situation, isn't it?

Reply:
By just considering the result as the same type we cannot handle situations like in one of the test cases:
```swift
func callitArgsFn<T>(_ : Int, _ f: () -> () -> T) -> T {
   f()()
}

callitArgsFn(1) {
     { print("hello") } // expected-error {{cannot convert value of type '()' to closure result type 'Int'}}
}
```
Where the result is `() -> T` but we hit the same problem, I didn't add a test case but the same may happen if result `(T, T)` maybe... But that is why I did go with contains

Reply:
It's effectively anything then right e.g. `(T) -> Void` as a result or `(Int, (String, T))` that's why I think it should be equals or maybe we should allow looking through function types too...

Reply:
Hum, but do you mean only look recursively through function type results?
And for tuples for example you mentioned we would have the same problem e.g. 
```swift
func callit<T>(_ : Int, _ f: () -> (T, Int)) -> T {
  f().0
}

func test() -> Int {
  callit(1) {
     (print("hello"), 0) // ambiguity 
  }
}
```

Reply:
Yes, recursively look at function results but nothing else because contextual mismatch between `callit` and `Int` is different from mismatch `(Void, Int)` and `(Int, Int)`.

Reply:
Ah I see thank you, so that would be a different problem right?